### PR TITLE
CloudWatch: Restrict auth provider and assume role usage according to Grafana configuration

### LINKED ISSUE #31082: AWS Plugins: Make it possible to enable/disable aws auth providers
### How it works now
In an AWS data source plugin, there are three authentication providers to choose from in the data source configuration page. These are *AWS SDK Default*, *Access & Secret Key* and *Credentials file*.

**AWS SDK Default** 
When this option is selected, no custom configuration is done in Grafana. Authentication is handled completely by the AWS Golang SDK. The SDK uses the first provider in the chain that returns credentials without an error. The default provider chain looks for credentials in the following order:
1. Environment variables.
2. Shared credentials file.
3. If your application uses an ECS task definition or RunTask API operation, IAM role for tasks.
4. If your application is running on an Amazon EC2 instance, IAM role for Amazon EC2.

There’s no way for Grafana to change this order of prioritization. 

**Access & Secret Key**
Uses the given access key ID and secret key to authenticate using the static credential provider in the AWS Golang SDK. This provider doesn’t have any fallbacks, and will fail if the provided key pair doesn’t work. 

**Credentials file**
Uses the shared credentials provider in the AWS Golang SDK. It will read the AWS shared credentials file and find the given profile. While AWS SDK Default will also find the shared credentials file, this option allows you to specify which profile to use without using environment variables. It doesn’t have any implicit fallbacks to other credential providers, and will fail if using credentials from the credentials file doesn’t work.

#### Assume role (ARN)
For each of the aforementioned authentication providers, it is possible to assume a role by using the *Assume Role* field in the data source config page. 

### Problem with the existing solution
* It doesn't make sense to have the *Credentials file* auth provider in a cloud hosted environment since the user can't access the server to upload a credentials file. 
* It doesn't make sense to have the *AWS SDK Default* auth provider in a cloud hosted environment since it's only the last option, *IAM role for Amazon EC2*, in the default provider chain that is relevant
* In Hosted Grafana it doesn't make sense to assume a role because it's not possible to edit permissions for the IAM role (if any) that is running the instance. [Question]: Will it be possible to Assume roles in AMG?

### Suggested Solution 
* For AMG and Hosted Grafana, hide *AWS SDK Default* and *Credentials file* options
* For AMG, introduce a new authentication option that will use *IAM role for Amazon EC2 authentication*. This option should be the default option for AMG. 
 [Question]: Which api in the golang SDK should we use to achieve that? 
 [Question]: What should we name this option? Suggestions are: *Workspace IAM role* and *Instance IAM role*

*OSS Grafana* 
Even though the fallback provider chain in *AWS SDK Default* can cause confusion to our users, we can't change that since it would potentially break authentication for some of our OSS users. We could potentially add *IAM role for Amazon EC2 authentication* not only for AMG, but also for OSS Grafana. However, it might make it more confusing since this is already possible to achieve using AWS SDK Default auth. 

### Implementation 
We can use [grafana.ini](https://grafana.com/docs/grafana/latest/administration/configuration/) to enable AWS authentication providers. It will be possible for external datasource plugins to access configuration in the grafana.ini as long as the fields are exposed in [config.ts](https://github.com/grafana/grafana/blob/master/packages/grafana-data/src/types/config.ts). 

It would be messy if each authentication provider had to be enable flag, so I suggest we introduce a variable representing a comma-separeted list of all the authentication providers in the grafana.ini.
```yaml
# Enter a comma-separated list of AWS authentication provider. Options are: aws_sdk_default, access_and_secret_key, credentials_file, ec2_IAM_role 
aws_auth_providers = aws_sdk_default,access_and_secret_key,credentials_file
```
OSS Grafana
```yaml
aws_auth_providers = aws_sdk_default,access_and_secret_key,credentials_file
```
AMG 
```yaml
aws_auth_providers = ec2_IAM_role,access_and_secret_key
```
Hosted Grafana 
```yaml
aws_auth_providers = access_and_secret_key
```

In the data source config, assume role (ARN) is not an authentication option per se, but more of an add-on feature to each provider. Therefore, there should be a separate flag for that in the grafana.ini. 
```yaml
## Allow AWS users to assume a role using temporary security credentials. If true, assume role will be enabled for all aws authentication providers that are specified in aws_auth_providers
aws_enable_assume_role =
```
Related to #25179

---

Does it sound like the right way forward for AMG and Hosted Grafana? @jongyllen @bergquist  @daniellee

Would this work for external AWS data source plugins? @ryantxu @aocenas

### PR DESCRIPTION
**Pull request by sunker**
**What this PR does / why we need it**:
This PR makes sure it's not possible to use an auth provider that is not specified as allowed in grafana.ini. Also making sure it's not possible to use assume roles if it's disabled in grafana.ini. For detailed information, see [this](https://github.com/grafana/grafana/issues/31082) issue. 
**Which issue(s) this PR fixes**:
Fixes #31082

### CODE REVIEW

**Review by aknuds1 — APPROVED** [review]
LGTM, but I think there might be a typo in the docs.

File: docs/sources/datasources/cloudwatch.md [hunk 1]
```diff
@@ -375,6 +375,18 @@ To request a quota increase, visit the [AWS Service Quotas console](https://cons
 
 Please see the AWS documentation for [Service Quotas](https://docs.aws.amazon.com/servicequotas/latest/userguide/intro.html) and [CloudWatch limits](https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/cloudwatch_limits.html) for more information.
 
+## Configure the data source with grafana.ini
+
+In the [Grafana configuration](https://grafana.com/docs/grafana/latest/administration/configuration/#aws) there's an `AWS` section that allows you to customize the data source.
+
+### allowed_auth_providers
+
+Specify what authentication providers should be allowed for the CloudWatch data source. The following providers are enabled by default in OSS Grafana: `default` (AWS SDK default), keys (Access and secret key), credentials (Credentials file), ec2_IAM_role (EC2 IAM role)
```
Comment (achatterjee-grafana) [inline docs/sources/datasources/cloudwatch.md:384]:
```suggestion
Specify which authentication providers are allowed for the CloudWatch data source. The following providers are enabled by default in OSS Grafana: `default` (AWS SDK default), keys (Access and secret key), credentials (Credentials file), ec2_IAM_role (EC2 IAM role).
```

**Review by achatterjee-grafana — CHANGES_REQUESTED** [review]
Requested one change.

File: pkg/tsdb/cloudwatch/cloudwatch.go [hunk 2]
```diff
@@ -64,19 +64,21 @@ func init() {
...
 func (s *CloudWatchService) NewExecutor(*models.DataSource) (plugins.DataPlugin, error) {
-	return newExecutor(s.LogsService), nil
+	return newExecutor(s.LogsService, s.Cfg), nil
 }
 
-func newExecutor(logsService *LogsService) *cloudWatchExecutor {
+func newExecutor(logsService *LogsService, settings *setting.Cfg) *cloudWatchExecutor {
 	return &cloudWatchExecutor{
 		logsService: logsService,
+		settings:    settings,
```
[the remark this answers is not in the document]
Comment (sunker) [reply pkg/tsdb/cloudwatch/cloudwatch.go]: Yeah I agree

**Review by achatterjee-grafana — APPROVED** [review]
LGTM

**Pull request merged**